### ISSUE TITLE
UploadFile: 425 Error while transfering data: ECONNABORTED - Connection aborted

### ISSUE THREAD

**Issue by FanDjango** [robinrodricks/FluentFTP#996]
**FTP OS:** Any

**FTP Server:** FileZille 1.5.x

**Computer OS:** WIndows 10

**FluentFTP Version:** V 41.1.0

**Framework:** .NET 6 AND NETFramework 4.8

```
			FTP_Sess.AutoConnect();

			FTP_Sess.UploadDirectory("D:\\temp\\issue311", "temp/issue311/upload", FtpFolderSyncMode.Update, FtpRemoteExists.Overwrite, FtpVerify.None, null, progress);

			FTP_Sess.Disconnect();
```

Files are uploaded, turn up fine on the server, but the transfer is reported as failed.

**Logs :**
```
[Error] GnuTLS error -110 in gnutls_record_recv: The TLS connection was non-properly terminated.
[Status] Client did not properly shut down TLS connection
[Trace] tls_layer_impl::shutdown()
[Trace] data_socket_->shutdown() = 126
[Response] 425 Error while transfering data: ECONNABORTED - Connection aborted
```

```
>         AutoConnect()
>         AutoDetect(True, False)
>         Connect()
Status:   FluentFTP 41.0.0.0
Status:   Connecting to ***:21
Status:   Waiting for a response
Response: 220-FileZilla Server 1.5.1
Response: 220-Please visit https://filezilla-project.org/
Response: 220 Welcome to AnSyNova FileZilla Server
Status:   Detected FTP server: FileZilla
Command:  AUTH TLS
Status:   Waiting for response to: AUTH TLS
Response: 234 Using authentication type TLS.
Warning:  SSL Buffering force disabled, is .NET 5.0 and later
Status:   FTPS authentication successful, protocol = Tls12
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0,1204242.
Command:  USER ***
Status:   Waiting for response to: USER ***
Response: 331 Please, specify the password.
Command:  PASS ***
Status:   Waiting for response to: PASS ***
Response: 230 Login successful.
Command:  PBSZ 0
Status:   Waiting for response to: PBSZ 0
Response: 200 PBSZ=0
Command:  PROT P
Status:   Waiting for response to: PROT P
Response: 200 Protection level set to P
Command:  FEAT
Status:   Waiting for response to: FEAT
Response: 211-Features:
Response: MDTM
Response: REST STREAM
Response: SIZE
Response: MLST type*;size*;modify*;perm*;
Response: MLSD
Response: AUTH SSL
Response: AUTH TLS
Response: PROT
Response: PBSZ
Response: UTF8
Response: TVFS
Response: EPSV
Response: EPRT
Response: MFMT
Response: 211 End
Status:   Text encoding: System.Text.UTF8Encoding+UTF8EncodingSealed
Command:  OPTS UTF8 ON
Status:   Waiting for response to: OPTS UTF8 ON
Response: 202 UTF8 mode is always enabled. No need to send this command
Command:  SYST
Status:   Waiting for response to: SYST
Response: 215 UNIX emulated by FileZilla.
Status:   Listing parser set to: Machine
>         UploadDirectory("D:\temp\issue311\from\", "temp/issue311/upload/", Update, Overwrite, None, null)
>         DirectoryExists("temp/issue311/upload")
Command:  PWD
Status:   Waiting for response to: PWD
Response: 257 "/" is current directory.
Command:  CWD temp/issue311/upload
Status:   Waiting for response to: CWD temp/issue311/upload
Response: 550 Couldn't open the file or directory
>         CreateDirectory("temp/issue311/upload", True)
>         DirectoryExists("temp/issue311")
Command:  PWD
Status:   Waiting for response to: PWD
Response: 257 "/" is current directory.
Command:  CWD temp/issue311
Status:   Waiting for response to: CWD temp/issue311
Response: 250 CWD command successful
Command:  CWD /
Status:   Waiting for response to: CWD /
Response: 250 CWD command successful
Status:   CreateDirectory temp/issue311/upload
Command:  MKD temp/issue311/upload
Status:   Waiting for response to: MKD temp/issue311/upload
Response: 257 "/temp/issue311/upload" created successfully.
>         UploadDirectoryFiles("3 files")
>         UploadFile("D:\temp\issue311\from\test1", "temp/issue311/upload/test1", NoCheck, False, None)
>         OpenWrite("temp/issue311/upload/test1", Binary)
Command:  TYPE I
Status:   Waiting for response to: TYPE I
Response: 200 Type set to I
>         OpenPassiveDataStream(PASV, "STOR temp/issue311/upload/test1", 0)
Command:  PASV
Status:   Waiting for response to: PASV
Response: 227 Entering Passive Mode (127,0,0,1,249,167)
Status:   Connecting to ***:63911
Command:  STOR temp/issue311/upload/test1
Status:   Waiting for response to: STOR temp/issue311/upload/test1
Response: 150 Starting data transfer.
Warning:  SSL Buffering force disabled, is .NET 5.0 and later
Status:   FTPS authentication successful, protocol = Tls12
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0,02694.
Status:   Disposing FtpSocketStream...
Status:   Waiting for response to: *UPLOAD*
Response: 425 Error while transfering data: ECONNABORTED - Connection aborted
Status:   Failed to upload file.
>         UploadFile("D:\temp\issue311\from\test2", "temp/issue311/upload/test2", NoCheck, False, None)
>         OpenWrite("temp/issue311/upload/test2", Binary)
>         OpenPassiveDataStream(PASV, "STOR temp/issue311/upload/test2", 0)
Command:  PASV
Status:   Waiting for response to: PASV
Response: 227 Entering Passive Mode (127,0,0,1,249,169)
Status:   Connecting to ***:63913
Command:  STOR temp/issue311/upload/test2
Status:   Waiting for response to: STOR temp/issue311/upload/test2
Response: 150 Starting data transfer.
Warning:  SSL Buffering force disabled, is .NET 5.0 and later
Status:   FTPS authentication successful, protocol = Tls12
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0,0074072.
Status:   Disposing FtpSocketStream...
Status:   Waiting for response to: *UPLOAD*
Response: 425 Error while transfering data: ECONNABORTED - Connection aborted
Status:   Failed to upload file.
>         UploadFile("D:\temp\issue311\from\test3", "temp/issue311/upload/test3", NoCheck, False, None)
>         OpenWrite("temp/issue311/upload/test3", Binary)
>         OpenPassiveDataStream(PASV, "STOR temp/issue311/upload/test3", 0)
Command:  PASV
Status:   Waiting for response to: PASV
Response: 227 Entering Passive Mode (127,0,0,1,249,171)
Status:   Connecting to ***:63915
Command:  STOR temp/issue311/upload/test3
Status:   Waiting for response to: STOR temp/issue311/upload/test3
Response: 150 Starting data transfer.
Warning:  SSL Buffering force disabled, is .NET 5.0 and later
Status:   FTPS authentication successful, protocol = Tls12
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0,0029047.
Status:   Disposing FtpSocketStream...
Status:   Waiting for response to: *UPLOAD*
Response: 425 Error while transfering data: ECONNABORTED - Connection aborted
Status:   Failed to upload file.
Command:  QUIT
Status:   Waiting for response to: QUIT
Response: 221 Goodbye.
Status:   Disposing FtpSocketStream...
```

**Comment by FanDjango** [comment]
@MunisoftEric
Move here, please, picking up #311

**Comment by FanDjango** [comment]
Server specific. Does not happen on proftpd and others.

**Comment by FanDjango** [comment]
vsftpd has a parameter to be less strict about the ssl_stream shutdown sequence, just for information.

**Comment by FanDjango** [comment]
Can be reduced to:
In FluentFTP 41.1.0
```
Command:  STOR temp/issue311/upload/test1
Status:   Waiting for response to: STOR temp/issue311/upload/test1
Response: 150 Starting data transfer.
Status:   FTPS authentication successful, protocol = Tls12
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0,0050078.
Status:   Disposing FtpSocketStream...
Status:   Waiting for response to: *UPLOAD*
Response: 425 Error while transfering data: ECONNABORTED - Connection aborted
Status:   Failed to upload file.
```
and in FileZilla 1.5.1
```
[Error] GnuTLS error -110 in gnutls_record_recv: The TLS connection was non-properly terminated.
[Status] Client did not properly shut down TLS connection
[Trace] tls_layer_impl::shutdown()
[Trace] data_socket_->shutdown() = 126
[Response] 425 Error while transfering data: ECONNABORTED - Connection aborted
```

**Comment by FanDjango** [comment]
Discovered some information about this:

[filezilla forum](https://forum.filezilla-project.org/viewtopic.php?f=2&t=7688)

This notice talks about the **server** handling TLS closure incorrectly. 

In our case, it is the **client** ( = the app that uses FluentFTP ), that is not closing correctly. The server problems appeared a long time ago, when the FileZilla client was being "improved". Out problem is quite new, as FileZilla Server was being similarly imporved. And other servers are following this path.

As with the current TLS 1.3 problems or the problems with "SSL Session Resume", this problem looks  like yet another weakness in Microsoft .NET.

More and more servers are adhering strictly to RFC and plugging security holes: 

- TLS **without** session resume is being **forbidden** for security reasons - there are many good articles on the internet about this and why that is actually good.

- TLS encrypted socket shutdown **without** a Closure Token sent by one of the parties would allow a spoofer to send a FIN paket and take over subsequent sessions as man in the middle. Using TLS and having this loophole open is viewed as being bad practice.

As usual, the open source community (FileZilla client and server at the forefront) are forcing these issues by programming this into their software without giving the user a way to "opt out". They say, that this is the only way to make the clients and servers out there follow suit and become more RFC compliant.

Well, we just want to do FTP **now** and sometimes one cannot roll back a server or a client to an earlier version.

FluenFTP uses .NET and for cryptography, the implementation of SslStream, which, in the end, transfers data through Windows' SChannel, which handles the low level secure data transfer.

FileZilla, on the other hand, used OpenSSL and after some problems, moved on to GnuTLS, even on windows.

Having to use SslStream means that:

Microsoft .NET pre WIN 11

* does not support TLS 1.3
* supports session resume on TLS 1.2, but we have seen already one problem with this on TLS 1.2, so it doesn't look perfect
* does not cleanly terminate an encrypted connection on TLS 1.2 (SslStream Dispose).

If you **are** using Win 11 and want TLS 1.3: It is turned off in FluentFTP for now as it is still not working reliably (Edge does it's own thing to a certain degree, it is not proof that .NET TLS 1.3 will work well).

**Comment by FanDjango** [comment]
I also found this: (from December 2016)

Closure Alert for SslStream:

This is required per [RFC 5246](https://tools.ietf.org/html/rfc5246#section-7.2.1) to avoid truncation attacks.

The SslStream included in .NET 4 does not adhere to this requirement. Somewhat surprisingly, it was decided by Microsoft to not fix the security problem, citing backwards compatibility as the reason.

**Comment by FanDjango** [comment]
I have taken a Wireshark trace to compare a download (whose shutdown is **not** faulted by FileZilla) and the upload (whose shutdown is faulted with the **425** error). I cannot see a Closure Alert on either. Only the upload causes a problem. On a download, an "incorrect" encryption shutdown is not rejected. That also explains why all other **incoming** encrypted data connections, such as when you are doing a file listing, are also ok, even if not "correctly" shut down. It seems FileZilla only cares about the connections uploading data to it.

**Comment by FanDjango** [comment]
I took another Wireshark trace, this time of WinSCP doing an upload with TLS 1.2, you can see:


In FileZilla server's log:
```
[Trace] tls_layer_impl::shutdown()
[Trace] tls_layer_impl::continue_shutdown()
[Trace] data_socket_->shutdown() = 0
[Response] 226 Operation successful
```
In Wireshark:
![image](https://user-images.githubusercontent.com/51046875/194701357-6ee4daf1-8c29-423a-b01c-5a5879136255.png)

This is the SSL shutdown sequence that is missing from SslStream dispose.

**Comment by FanDjango** [comment]
Here is some intriguing information that contradicts some of what I found and described earler:

**Comment by vasyaxparfenov** [comment]
@FanDjango Did you provide config value for closing ssl stream? FtpConfig.DisconnectWithShutdown
You're right about older versions of framework though. They don't support SslShutdown

**Comment by FanDjango** [comment]
I have collected all open issues that are caused by "Incorrect SSL shutdown" here:

#771, #311, #930, #573, #944, #763.

The error has been observed on newer FileZilla vsftpd FTP servers and on a Unisys FTP server.

History:

This problem had already been **partly** solved. There was `FtpSslStream` that performed the same job as my `FluentSslStream` wrapper (see #997). It used the same technology to issue the Closure Alert on the SslStream.

**But:** It did not support .NET 6. It only supported .NET Framework. So it was only a part solution anyway.

**And then:** Even for .NET Framework, it didn't kick in because of unfortunate wrong order in disposing the streams. This was subsequently fixed, released only a few days ago.

**But, but, but:** The `FtpSslStream` is no longer there, it is no longer being used - it was removed sometime around beginning of September.

Happily, the wrapper #997 is now merged, always active in all targets, doesn't check `can_write`, supports **both** .NET 6.0 **and** .NET Framework, so things should get better as soon as this is released, probably as V 41.1? @robinrodricks ?

**Comment by FanDjango** [comment]
@vasyaxparfenov 
>  FluentFTP previously had logic for doing SSL Shutdown but due to incorrect calls order in FtpSocketStream.Dispose it never got called even if config was set.

Total agreement. Took me a while researching all the old issues, the code and so on to see that.

**Comment by FanDjango** [comment]
@vasyaxparfenov 

> Did you provide config value for closing ssl stream? FtpConfig.DisconnectWithShutdown

No. It was decided to shutdown correctly, period.

**Comment by FanDjango** [comment]
@vasyaxparfenov 
> Did you provide config value for closing ssl stream? FtpConfig.DisconnectWithShutdown

Good catch - checking that right now.

**Comment by FanDjango** [comment]
It looks like setting `Config.DisconnectWithShutdown = true` works for .NET 6.0, even without the wrapper. It causes `ShutdownAsync()` to be used, in the few cases where it is available.

I could imagine moving that code into the wrapper and to remove the config parameter ( = always true ).

**Comment by vasyaxparfenov** [comment]
I would say it correctly Shutdowns in case config provided for NET5_0_OR_GREATER || NETSTANDARD2_1_OR_GREATER which should cover .NET Core 3.1 also. Not sure whether using wrapper with reflection is good for everybody. Maybe we could add it as a workround for older version with notes on implementation details and possible performance costs due to reflection. I think older versions of .NET framework are not going to be supported forever

**Comment by vasyaxparfenov** [comment]
Thing is SslStream is known issue of .NET Framework so it is actually beyond our reach. I would ask consumers to upgrade to .net standard 2.1 compliant versions in case they want latests security patches since all latets versions of FTP servers demand this behaviour.

**Comment by FanDjango** [comment]
Thanks for your input. I am just searching for the way to get this running again even for some of the less wonderful targets.

> I would say it correctly Shutdowns in case config provided for NET5_0_OR_GREATER || NETSTANDARD2_1_OR_GREATER which should cover .NET Core 3.1 also

Yes, I just tested that. `AsyncShutdown()` FTW. Perhaps one should just change the default of `Config.DisconnectWithShutdown` to true, as one measure. **Not sure**, by the way, in all the issues listed above, that this parameter has been set, at all. 

Didn't really need all the effort I put in to make the wrapper work for NET5_0_OR_GREATER, in that case. I could just remove that part again, and modify `CreateSslStream` to be like it used to be - depending on NET5_0_OR_GREATER.

> possible performance costs due to reflection

For a dispose once in a while... but yeah, if they are transferring directories with thousands of small files.

> Otherwise, we would probably need to implement custom SslStream

That monster is lurking somewhere on the horizon. Let's hope it stays there.

**Comment by vasyaxparfenov** [comment]
Alternatively, we can just handle error returned from server. At the end of day, tls truncation technically is aimed against servers. In this case modern ftps servers are aware of it so in case TCP FIN arrives before SslShutdown - they will react, which means client should also react. You can't trick client otherwise, since if response is not sent from server to client - client would throw exception saying connection terminated so our consumers are aware. We can add this to documentation as I said and specifically point out case for using FTP Client for older frameworks. Please, correct me If I'm wrong

**Comment by FanDjango** [comment]
Yes. It actually really means dropping RFC compliant secure support for uploads in all of .NET Framework and other older stuff. If they insist, then on a modern server, they will get a 226.... message. Some servers will keep the file, some will remove it. That's where a bit of a problem lies, one cannot rely on all servers behaving the same.

Am I right in reading your post that it more or less says: **Let's actually not fix it for .NET Framework?**

**Comment by FanDjango** [comment]
Anyway, with the current **master** the **NET5_0_OR_GREATER || NETSTANDARD2_1_OR_GREATER** Closure Alert will be sent by `AsyncShutdown()`.

Otherwise: The SSL wrapper will be used (reflection, unsafe, marshalling etc.).

Closure Alert will **always** be done, regardless of the setting of `Config.DisconnectWithShutdown`. It no longer controls that.

The `Config.DisconnectWithShutdown` parameter **still** controls wether the socket is asked to shutdown prior to its disposal. The question arises, why not **always** do that on the socket also and remove the parameter entirely.

**Comment by vasyaxparfenov** [comment]
@FanDjango actually I wasn’t right. SSL Close Alert is more for client to verify whether connection hasn’t been closed by MIM. I think an idea with wrapper for Framework and current implantation with implicit Shutdown for versions compliant with Standard 2.1 is a good way to go! Only reason why I wouldn’t support Framework is wrapper should be attentively supported in case any vulnerabilities would need to be fixed in future for Close alert and by dropping frameworks support in this matter might push engineers to upgrade)))

**Comment by FanDjango** [comment]
> wrapper for Framework and current implementation with implicit Shutdown for versions compliant with Standard 2.1 is a good way to go

Yes, am moving in that direction

What is your opinion on 

> The Config.DisconnectWithShutdown parameter still controls wether the socket is asked to shutdown prior to its disposal. The question arises, why not always do that on the socket also and remove the parameter entirely.

**Comment by FanDjango** [comment]
@vasyaxparfenov I wonder why servers only complain on uploads. On downloads, they seem to be more lenient

**Comment by vasyaxparfenov** [comment]
@FanDjango my opinion on removing config and making it default behavior (implicit) is same as yours - positive. Regarding download vs upload - my guess is when we upload by closing data connection we imply that we finished, it’s like a signal for server, when we download I suppose a server should be one closing data and signaling us hence it doesn’t fail when downloading because server should be closing ssl and subsequently tcp connection.

**Comment by FanDjango** [comment]
@vasyaxparfenov Sort of like saying: He who is sending, he should say "OVER AND OUT", otherwise opening up a security issue for MIM just going on and on.

**Comment by vasyaxparfenov** [comment]
@FanDjango Yeah, that’s my guess. You can check in Wireshark though

**Comment by FanDjango** [comment]
For this issue and the needed changes I have noticed the following: If in the previous version of the code you asked for `DisconnectWithShutdown`: the SSL Closure Alert would be sent and then the socket would be given a shutdown call.

Nobody seems to have noticed this, but this actually breaks the timeout logic - our integration tests (timeout) fail once you do this. It's just that nobody who used this parameter got into the situation where he would need to have the timeouts work.

**Comment by vasyaxparfenov** [comment]
@FanDjango I guess issue was with CloseSocket logic. It seemed to send [Shutdown](https://github.com/robinrodricks/FluentFTP/commit/0af34c1b075940cb551a915c56ac844c792c0982) to socket which might have not been fully initiated due to timeout. However, you seemed to remove that bit already.

**Comment by FanDjango** [comment]
@vasyaxparfenov 

>  However, you seemed to remove that bit already

Yes, a short time ago I did that. The CloseSocket logic is now back to the basics, so to speak. The only thing worrying me is the slight differences in code re. the targets but any glitches must be ironed out in the future. I would appreciate your further input on this in the next time.

> So we might even consider removing extra bits added on top of logic and just pass cancellation token which fires after timeout value.

I will need to research and analyze to fully understand that sentence*, but I have an inkling what you are getting at. It is just important to keep all edge cases in mind that have been there in the past and that led to some of the pieces of code that became redundant with later improvements in other areas.

*) Probably not the highest priority right now.

**Comment by vasyaxparfenov** [comment]
@FanDjango it is actually very simple with Timeout. We register cancellation token action which calls our CloseSocket logic. Out CloseSocket logic doesn't know about internals so, in case DisconnectWithShutdown is true it just sends Shutdown to never fully initiated Socket - this is wrong. We just need to use [this](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.connectasync?view=net-6.0#system-net-sockets-socket-connectasync(system-string-system-int32-system-threading-cancellationtoken)) overload and provide it our [cancellation token](https://github.com/robinrodricks/FluentFTP/blob/0af34c1b075940cb551a915c56ac844c792c0982/FluentFTP/Streams/FtpSocketStream.cs#L954) without any action and overload will handle it in a correct way.

**Comment by FanDjango** [comment]
@vasyaxparfenov Do you mind if I copy out this conversation of ours, archive it in my todo list and then delete the posts, so to keep this thread totally on topic for the actual "Incorrect SSL Shutdown" topic? I would open a separate issue for a conversation about how to further improve this part of the code with you.

**Comment by vasyaxparfenov** [comment]
@FanDjango sure, thank you!

**Comment by FanDjango** [comment]
Reminder and summary of the current status:

Until everyone gets the upcoming V42 of FluentFTP, which should fix this, 

those of you who are using NET 5.0 or NET 6.0:
**Set Config.DisconnectWithShutdown = true.**

those of you using .NET Framework:
**wait for V42**

**Comment by FanDjango** [comment]
V42 is released as of today, https://github.com/robinrodricks/FluentFTP/issues/771, https://github.com/robinrodricks/FluentFTP/issues/311, https://github.com/robinrodricks/FluentFTP/issues/930, https://github.com/robinrodricks/FluentFTP/issues/573, https://github.com/robinrodricks/FluentFTP/pull/944, https://github.com/robinrodricks/FluentFTP/issues/763 please take note

**Comment by robinrodricks** [comment]
Closing as @FanDjango  has mentioned this is fixed.
